libs/MotionKit/include/MotionKit.hpp

```diff
@@ -5,65 +5,42 @@
 #pragma once
 
 #include <interface/drivers/Motor.h>
-#include <interface/libs/EventLoop.h>
 
-#include "IMUKit.hpp"
-#include "PID.hpp"
+#include "RotationControl.hpp"
 #include "interface/drivers/Timeout.h"
+#include "interface/libs/IMUKit.hpp"
 
 namespace leka {
 
 class MotionKit
 {
   public:
-	MotionKit(interface::Motor &motor_left, interface::Motor &motor_right, IMUKit &imu_kit,
-			  interface::EventLoop &event_loop, interface::Timeout &timeout)
-		: _motor_left(motor_left),
-		  _motor_right(motor_right),
-		  _imukit(imu_kit),
-		  _event_loop(event_loop),
-		  _timeout(timeout)
+	MotionKit(interface::Motor &motor_left, interface::Motor &motor_right, interface::IMUKit &imu_kit,
+			  interface::Timeout &timeout)
```

Comment:
in the future, MotionKit will **not** depend on IMUKit -- the controls will depend on what they need, including IMUKit, Motors if needed

Controls will be like "dynamic" motion behaviors.

when a control starts, it will register the needed callback to IMUKit, replacing the old one.

when stopping from MotionKit, the callback is to be removed.

Reply:
> in the future, MotionKit will not depend on IMUKit -- the controls will depend on what they need, including IMUKit, Motors if needed

I know that MotionKit should not depend on IMUKit and I agree with that. Now consider MotionKit is responsible of Motion of the robot, what ensure that when MotionKit is asked to be stopped the robot effectively stopped? What if a Control still calling in IMUKit callback a new speed in motors (we forget to remove the callback at stop in the Control file for instance)?

Since it is not a problematic in this PR, we can resolve this conversation here, just keep it in mind when the next refactor of MotionKit will come.

Reply:
> Now consider MotionKit is responsible of Motion of the robot, what ensure that when MotionKit is asked to be stopped the robot effectively stopped? What if a Control still calling in IMUKit callback a new speed in motors (we forget to remove the callback at stop in the Control file for instance)?

Welcome to callback hell! :)

This is the kind of issue that can arise when callbacks are not used well.

It's true for a callback but it's also true for other dependencies.

Lets take an example with `ExampleControl`: (it's an example, some things might not exist yet/at all)

- it depends on IMUKit for sensor data
- it depends on two motors for motion
- it also depends on LedKit so also change the color while moving
- it depends on BatteryKit to change the intensity and motor speed depending of the battery level

There's no reason for MotionKit to depend on all those things to make sure that:

- IMUKit callback as been set to an empty std::function
- the two motors are really stopped
- leds are turned off
- BatteryKit callback has also been removed

it's a bit like **not using** Smart Pointers: if you use raw C pointers and , you need to make sure you clean up after yourself.

Following the TDD methodology, this is what we should be doing:

- specify how `ExampleControl` will work
- specify it's dependencies
- review specifications as a team to spot this kind of issues and add a specific task or checklist to make sure we don't forget
- write the tests needed to enforce that on `stop()` the dependencies are really cleared/turned off
- have the tests fail
- implement `stop()` to pass the tests

There's no simple way to do that "automatically" without complex engineering.

it's something that might be possible to do with concepts and templates. Surely an amazing engineering experience but I'm not sure it's worth the time and effort right now.

Again, we master all our code. We don't depend on something opaque and don't know what it does, nor do we provide these libraries as public API for outside users.